### ISSUE TITLE
#4: Cadence Scenario: Client API

### ISSUE THREAD

**Issue by jefflill** [nforgeio/neonKUBE#513]
### Note to JACK:

- [x] **BREAKING:** I've renumbered renumbering the message IDs and also removed some that we're not going to use.  You'll need to update these definitions for the cadence-proxy to match.

- [x] **BREAKING:** I've renamed both the `WorkflowContextId` property for workflow request messages and the `ActivityContextId` for activity requests to `ContextId`.  I did this because the old way implied that that workflow and activity contexts were distinct, but these are really the same thing.  You'll continue to maintain a single dictionary of contexts that will be used for both workflows and activities.

- [x] `WorkflowInvokeReply:` I've added these several new properties: `bool ContinueAsNew` and `byte[] ContinueAsNewArgs` as well as some workflow context options (described below).  These are designed to support the Cadence continue-as-new functionality.  This feature allows a workflow to have itself restarted with the same ID and also with a clean history.

  This use case will be probably pretty common for Loopie.  This will be useful for looping workflows that may run for a very long time.  For example, a washer workflow that runs once a day summarizes the washer activities and submits their pay to PayPal.  This could be implemented using a loop in the workflow code, but this would mean the task history would grow for every loop, which would be unnecessary and in-efficient.

  Here's how you'll handle this:

  1. If `WorkflowInvokeReply.ContinueAsNew == false` then you'll do what you're already doing: signal the workflow future to return the result/error returned in the message.

  2. If `WorkflowInvokeReply.ContinueAsNew == true` then you'll need to call `NewContinueAsNewError()` and then have your workflow future return this error to Cadence.  The tricky part is that you may need to call various `context.WithXXX()` methods to add options to the context before calling `NewContinueAsNewError()`.  This will look something like:

  **Note to Jack:** I removed the retry policy related code below since there doesn't appear to be a way to set it.  I don't believe this is a big loss for this scenario.

  ```
  if (reply.ContinueAsNew) {
  
    continueContext := context;
	
	if (reply.ContinueAsNewDomain != null) {
	  continueContext = continueContext.WithDomain(reply.ContinueAsNewDomain);
	}
	
    if (reply.ContinueAsNewTaskList != null) {
	  continueContext = continueContext.WithTaskList(reply.ContinueAsNewTaskList);
	}
	
    if (reply.ContinueAsNewExecutionStartToCloseTimeout > 0) {
      continueContext = continueContext.WithExecutionStartToCloseTimeout(reply.ContinueAsNewExecutionStartToCloseTimeout );
    }
	
	if (reply.ContinueAsNewScheduleToCloseTimeout > 0) {
      continueContext = continueContext.WithScheduleToCloseTimeout(reply.ContinueAsNewScheduleToCloseTimeout );
    }  
	
	if (reply.ContinueAsNewScheduleToStartTimeout > 0) {
      continueContext = continueContext.WithScheduleToStartTimeout(reply.ContinueAsNewScheduleToStartTimeout );
    }  
	
	if (reply.ContinueAsNewStartToCloseTimeout > 0) {
      continueContext = continueContext.WithStartToCloseTimeout(reply.ContinueAsNewStartToCloseTimeout );
    }
	
	workflowFunction := // create a new workflow function here.
	
	continueError := NewContinueAsNewError(continueContext, workflowFunction, reply.ContinueAsNewArgs);
	
	// Set [continueError] on your workflow future so that your workflow function
	// will return this error back to Cadence.
  }
  ```

- [x] `WorkflowReply:` I neglected to rename the `WorkflowContextid` property to `ContextId`.  I just renamed this now and merged it to the **cadence** branch.

- [x] **BREAKING:** I've removed the `ProxyRequest.Timeout` property because I handle timeouts on the client side.

---

The objective for this scenario is to start laying out the .NET client API and performing related operations end-to-end.

### cadence-client prerequisites

- [x] Enhance `IRetryPolicy` and its implementations to support an optional maximum retry interval that overrides the other policy variables when exceeded.  This will align completely with Cadence retry policies.

**New Messages:**

- [x] `PingRequest`
- [x] `PingReply`
- [x] `WorkflowCancelRequest`
- [x] `WorkflowCancelReply`
- [x] `WorkflowTerminatRequest`
- [x] `WorkflowTerminateReply`
- [x] `WorkflowSignalRequest`
- [x] `WorkflowSignalReply`
- [x] `WorkflowSignalWithStartRequest`
- [x] `WorkflowSignalWithStartReply`
- [x] `WorkflowQueryRequest`
- [x] `WorkflowQueryReply`
- [x] `WorkflowSetCacheSizeRequest`
- [x] `WorkflowSetCacheSizeReply`
- [x] `WorkflowInvokeRequest` **needs echo testing**
- [x] `WorkflowInvokeReply`  **needs echo testing**
- [x] `WorkflowMutableRequest`
- [x] `WorkflowMutableReply`
- [x] `WorkflowMutableInvokeRequest`
- [x] `WorkflowMutableInvokeReply`
- [x] `WorkflowDescribeExecutionRequest`
- [x] `WorkflowDescribeExecutionReply`
- [x] `WorkflowGetResultRequest`
- [x] `WorkflowGetResultReply`
- [x] `WorkflowSignalSubscribeRequest`
- [x] `WorkflowSignalSubscribeReply`
- [x] `WorkflowSignalReceivedRequest`
- [x] `WorkflowSignalReceivedReply`
- [x] `WorkflowHasLastResultRequest`
- [x] `WorkflowHasLastResultReply`
- [x] `WorkflowGetLastResultRequest`
- [x] `WorkflowGetLastResultReply`
- [x] `ActivityExecuteRequest`
- [x] `ActivityExecuteReply`
- [x] `ActivityInvokeRequest`
- [x] `ActivityInvokeReply`
- [x] `ActivityGetHeartbeatDetailsRequest`
- [x] `ActivityGetHeartbeatDetailsReply`
- [x] `ActivityRecordHeartbeatRequest`
- [x] `ActivityRecordHeartbeatReply`
- [x] `ActivityHasHeartbeatDetailsRequest`
- [x] `ActivityHasHeartbeatDetailsReply`
- [x] `ActivityStoppingRequest`
- [x] `ActivityStoppingReply`
- [x] `WorkflowDisconnectContextRequest`
- [x] `WorkflowDisconnectContextReply`
- [x] `WorkflowGetTimeRequest`
- [x] `WorkflowGetTimeReply`
- [x] `WorkflowSleepRequest`
- [x] `WorkflowSleepReply`
- [x] `WorkflowExecuteChildRequest`
- [x] `WorkflowExecuteChildReply`
- [x] `WorkflowExecuteChildReply` **REFACTORED:** See instructions for **Refactor Child Execute** below.
- [x] `WorkflowWaitForChildRequest`
- [x] `WorkflowWaitForChildReply`
- [x] `WorkflowSignalChildRequest`
- [x] `WorkflowSignalChildReply`
- [x] `WorkflowCancelChildRequest`
- [x] `WorkflowCancelChildReply`
- [x] `WorkflowSetSignalHandlerRequest`
- [x] `WorkflowSetSignalHandlerReply`
- [x] `WorkflowSetQueryHandlerRequest`
- [x] `WorkflowSetQueryHandlerReply`
- [x] `ActivityRegisterRequest`
- [x] `ActivityRegisterReply`
- [x] `ActivityExecuteLocalRequest`
- [x] `ActivityExecuteLocalReply`
- [x] `ActivityInvokeLocalRequest`
- [x] `ActivityInvokeLocalReply`
- [x] `WorkflowQueryInvokeRequest`
- [x] `WorkflowQueryInvokeReply`
- [x] `ActivityGetInfoRequest`
- [x] `ActivityGetInfoReply`
- [x] `ActivityCompleteRequest`
- [x] `ActivityCompleteReply`

### cadency-proxy prerequisites:

- [x] `PingRequest`
- [x] `PingReply`
- [x] `WorkflowCancelRequest` **note:** added the **Domain** property.
- [x] `WorkflowCancelReply`
- [x] `WorkflowTerminateRequest`
- [x] `WorkflowTerminateReply`
- [x] `WorkflowSignalRequest`
- [x] `WorkflowSignalReply`
- [x] `WorkflowSignalWithStartRequest` **note:** added the **Workflow** property to hold the workflow name.
- [x] `WorkflowSignalWithStartReply` **note:** added the **ID** and **ReplyID** properties.
- [x] `WorkflowQueryRequest`
- [x] `WorkflowQueryReply`
- [x] `WorkflowSetCacheSizeRequest`
- [x] `WorkflowSetCacheSizeReply`
- [x] `ConnectRequest` **note:** Added the **ClientTimeout** property.
- [x] `WorkflowInvokeRequest` **needs echo testing**
- [x] `WorkflowInvokeReply` **needs echo testing**
- [x] `WorkflowMutableRequest`
- [x] `WorkflowMutableReply`
- [x] `WorkflowMutableInvokeRequest`
- [x] `WorkflowMutableInvokeReply`
- [x] `WorkflowDescribeExecutionRequest`
- [x] `WorkflowDescribeExecutionReply`
- [x] `WorkflowGetResultRequest`
- [x] `WorkflowGetResultReply`
- [x] `WorkflowSignalSubscribeRequest`
- [x] `WorkflowSignalSubscribeReply`
- [x] `WorkflowSignalReceivedRequest`
- [x] `WorkflowSignalReceivedReply`
- [x] `WorkflowHasLastResultRequest`
- [x] `WorkflowHasLastResultReply`
- [x] `WorkflowGetLastResultRequest`
- [x] `WorkflowGetLastResultReply`
- [x] `ActivityExecuteRequest`
- [x] `ActivityExecuteReply`
- [x] `ActivityInvokeRequest`
- [x] `ActivityInvokeReply`
- [x] `ActivityGetHeartbeatDetailsRequest`
- [x] `ActivityGetHeartbeatDetailsReply`
- [x] `ActivityRecordHeartbeatRequest`
- [x] `ActivityRecordHeartbeatReply`
- [x] `ActivityHasHeartbeatDetailsRequest`
- [x] `ActivityHasHeartbeatDetailsReply`
- [x] `ActivityStoppingRequest`
- [x] `ActivityStoppingReply`
- [x] `WorkflowDisconnectContextRequest`
- [x] `WorkflowDisconnectContextReply`
- [x] `WorkflowGetTimeRequest`
- [x] `WorkflowGetTimeReply`
- [x] `WorkflowSleepRequest`
- [x] `WorkflowSleepReply`
- [x] `WorkflowExecuteChildRequest`
- [x] `WorkflowExecuteChildReply`
- [ ] `WorkflowExecuteChildReply` **REFACTORED:** See instructions for **Refactor Child Execute** below.
- [x] `WorkflowWaitForChildRequest`
- [x] `WorkflowWaitForChildReply`
- [x] `WorkflowSignalChildRequest`
- [x] `WorkflowSignalChildReply`
- [x] `WorkflowCancelChildRequest`
- [x] `WorkflowCancelChildReply`
- [x] `WorkflowSetSignalHandlerRequest`
- [x] `WorkflowSetSignalHandlerReply`
- [x] `WorkflowSetQueryHandlerRequest`
- [x] `WorkflowSetQueryHandlerReply`
- [x] `ActivityRegisterRequest`
- [x] `ActivityRegisterReply`
- [x] `ActivityExecuteLocalRequest`
- [x] `ActivityExecuteLocalReply`
- [x] `ActivityInvokeLocalRequest`
- [x] `ActivityInvokeLocalReply`
- [x] `WorkflowQueryInvokeRequest`
- [x] `WorkflowQueryInvokeReply`
- [x] `ActivityGetInfoRequest`
- [x] `ActivityGetInfoReply`
- [x] `ActivityCompleteRequest`
- [x] `ActivityCompleteReply`

### Cadence-Proxy Implementation

- [x] `PingRequest/PingReply` are intended for testing the performance of performing transactions between the **cadence-client** and the **cadence-proxy**.  These are very simple messages that may originate from either the **cadence-client** or **cadence-proxy** side, with the receiving side simply responding to a `PingRequest` with a `PingReply`.

- [x] `CancelWorkflowRequest:` Call the `Client.CancelWorkflow() function, passing the workflow ID and RunID passed in the message.

- [x] `TerminateWorkflowRequest:` Call the `Client.TerminateWorkflow()` function, passing the workflow ID and RunID passed in the message, along with the reason string and details.

- [x] `SignalWorkflowRequest:` Call the `Client.SignalWorkflow()` function, passing the message parameters.

- [x] `SignalWorkflowWithStartRequest:` Call the `Client.SignalWithStartWorkflow()` function, passing the the message parameters.

- [x] `WorkflowQueryRequest:` Call the `Client.QueryWorkflow()` function, passing the message parameters.

- [x] `SetWorkflowCacheSizeRequest:` Call the `worker.SetStickyWorkflowCacheSize()` function, passing the size parameter from the message.

- [x] The .NET related workflow messages have been refactored a bit.  I've renamed `WorkflowContextRequest --> WorkflowRequest` and `WorkflowContextReply` and modified all workflow related requests to derive from `WorkflowRequest` and all replies from `WorkflowReply`.  Note that not all of these messages will need to set the `WorkflowContextId` properties, but I thought the consistency would be nicer.

- [x] I've added the following properties to the `WorkflowInvokeRequest` message:

  * `Domain`
  * `WorkflowId`
  * `RunId`
  * `WorkflowType`
  * `TaskList`
  * `ExecutionStartToCloseTimeout`
  &nbsp;
  You'll need to call `context.GetWorkflowInfo()` before sending this message to me to obtain these values.
&nbsp;
- [x] `WorkflowMutableRequest:` This one is a double-reverse tricky.  This is intended to implement the Cadence side effect functionality.  The basic use case is introduce a named value that can be retrived indeterminately while running a workflow.  From the .NET app perspective, I'll have a method in the base `Workflow` class like:
  ```
  public class Workflow
  {
     protected async Task<byte[]> GetMutableValueAsync(string mutableId, Func<Task<byte[]>> getter);
  }
  ```
  The semantics are:

  1. The first time this is called on a workflow run and for a specific `mutableId`, the `getter` function will be called to retrieve the actual result.  Cadence will persist this result using the `mutableId` and then return the result to the workflow.

  2. For subsequent calls on a workflow run for the same `mutableId`, Cadence won't call the the function and will instead return the value it result it saved previously.

  Here's how this will be implemented:

  1. The .NET `Workflow.GetMutableValueAsync()` method will send a `WorkflowMutableRequest` to the cadence-proxy, passing the `WorkflowContextId` and `MutableId` string.

  2. cadence-proxy will call `workflow.MutableSideEffect()` passing it the workflow context (retrieved via the `workflowContextId`) and a function that does the following:

  3. The function will send a `WorkflowInvokeMutableRequest` message back to the cadence-client, passing the `WorkflowContextId` and `MutableId`.  The function will then wait for a `WorkflowInvokeMutableReply` from the cadecne client.
	
  4. The `WorkflowInvokeMutableReply` will return either a `Result` or `Error`.  You'll return these from your function, back to Cadence.
	
  5. The code that waiting on the `workflow.MutableSideEffect()` call will receive the result and/or error and package these into a `WorkflowMutableReply` and send that back to the cadence-client.
  
  6. The cadence-client will receive this and return the result back to the workflow as the result of the original `GetMutableValueAsync()` method or throw an exception.

- [x] `WorkflowDescribeExecutionRequest:` Call the corresponding client function.  The result should be able to be serialized to JSON and then be deserialized on the .NET side into the `InternalDescribeWorkflowExecutionResponse` class.  Let me know if this doesn't work (there were a lot of properties and subtypes).

- [x] `WorkflowGetResultRequest:` cadence-proxy will get a `WorkflowRun` via the `client.GetWorkflow()` function, call its `Get()` method and return the result/error back to the cadence-client via a `WorkflowGetResultReply`.

- [x] `WorkflowSignalSubscribeRequest:` The cadence-proxy will need to get the named signal channel and register a receive function something like shown [here](https://cadenceworkflow.io/docs/03_goclient/08_signals).  The receive function will send the received signal name and payload to the cadence-client via a `WorkflowSignalReceivedRequest` message and wait for the corresponding reply.

- [x] I've added the `Memo` property to `InternalStartWorkflowOptions` and `InternalChildWorkflowOptions`.  This change should be transparent to the cadence-proxy.

- [x] `WorkflowHasLastResultRequest:` You'll need to call the `client.HasLastCompletionResult()` function.

- [x] `WorkflowGetLastResultRequest:` You'll need to call the `client.GetLastCompletionResult()` method.

- [x] `ActivityExecuteRequest:` You'll need to lookup the workflow context via `WorkflowContextId`, use `WithActivityOptions()` to set the `Options` and then call `context.ExecuteActivity()`, passing the activity arguments and a custom activity execution function.  `context.ExecuteActivity()` returns a future so you'll send the `ActivityExecuteReply` back to the cadence-client when it's `Get()` method returns a result or error.

  Your activity execution function will work much like how we handled workflow execution: you'll send an `ActivityInvokeRequest` to me and then wait on a future for my `ActivityInvokeReply` with the result/error.

  **IMPORTANT:** You should never persist the new context returned by `context.WithActivityOptions()` or any of the other `WithXXX()` calls back to your workflow context dictionary for this or **any other message implementations**.  The idea here is that those contexts need to remain unchanged and we'll always use the `WithXXX()` methods (if necessary) to customize the context before performing any operations on it.

- [x] `ActivityGetHeartbeatDetailsRequest:` Call `context.GetHeartbeatDetails()` on the context and respond with an `ActivityGetHeartbeatDetailsReply` with the result/error.

- [x] `ActivityRecordHeartbeatRequest:` Lookup the context and then call `activity.RecordActivityContext()`, passing the details.

- [x] `ActivityHasHeartbeatDetailsRequest:` Lookup the context and then call `activity.HasHeartbeatDetails()` and send the result back to the client.

- [x] `ActivityStoppingRequest:` This message will be sent by the cadency-proxy to the cadence-client to indicate that the the worker running the activity is stopping, giving the activity a chance to terminate cleanly (e.g. send a last heartbeat with progress info).  This message adds an activity id field.

  @johncburns1: You'll need to modify your activity registration code:

  The function you pass to `RegisterActivity()` should call `context.GetWorkerStopChannel()` on the context passed to the function and then start a go routine that waits for the stop channel to be closed.  When this happens, you need to send a `ActivityStoppingRequest` to the client (with the activity ID) and wait for the `ActivityStoppingReply`.  The client will use this to signal the .NET application's activity to stop.

- [x] Disable compression the HTTP client the cadence -proxy uses to submit requests to the cadence-client if possible.  There's not a lot of point burning CPU doing compression when this communication is essentially an in-memory operation.  I've disabled this when the cadence-client submits requests to the proxy.  Let me know if this isn't possible and I can probably rig my web server to disable this.

- [x] `WorkflowDisconnectContextRequest:` You'll need to lookup you context record and the call `context.NewDisconnectedContext()` to retrieve the replacement context.  You'll then replace the existing workflow context with this new value.

- [x] `WorkflowGetTimeRequest:` You'll need to call `context.Now()` and return the time in `WorkflowContextReply.Time`.  I'm expecting this to be UTC.

- [x] `WorkflowSleepRequest:` You'll need to call `context.Sleep()`, passing the duration and then send the `WorkflowSleepReply` after the `context.Sleep()` method returns.

- [x] `WorkflowExecuteChildRequest:` You'll lookup the context, call `context.WithChildOptions()` passing the `Options` from the message and then use the context returned to call `ExecuteChildWorkflow()`, passing the `Args` property as the workflow arguments.  `ExecuteChildWorkflow()` returns a future.  You'll need to wait on that and send a `WorkflowExecuteChildReply` back to me with the result/error.

- [x] **Refactor Child Execute:**

  You're going to need to refactor how executing child workflows works, as we discussed.  I've split this into two operations `WorkflowExecuteChildRequest/Reply` which will now start the child workflow **but not wait for it to complete**.  You'll need to add a table to your context struct that maps long child IDs to a new `ChildContext` struct (or something) that holds the `ChildWorkflowFuture` returned by the `ExecuteChildWorkflow()` function as well as a cancellation function.
  
  You'll need to generate a unique `childId` in a threadsafe way, like you're already doing for context IDs (technically, `childId` is scoped to the parent workflow, but having this be global is fine).  `WorkflowExecuteChildRequest` should create a cancellation function and pass it to `WithCancel()`, start the new workflow, and then create the local child context, setting the future returned as well as the cancellation function, and then send the `WorkflowExecuteChildReply` back to me, including the `childId`.
  
- [x] `WorkflowWaitForChildRequest:` You'll locate the parent workflow context and then use the `contextId` included in the message to locate the child context record.  If this doesn't exist, you should immediately send the reply with an **EntityNotExists** error.  Otherwise, you'll call `Get()` on the future, remove the client context and then send the result back to the client in the reply message.

- [x] `WorkflowSignalChildRequest:` You'll locate the parent workflow context and then use the `contextId` included in the message to locate the child context record.  If this doesn't exist, you should immediately send the reply with an **EntityNotExists** error.  Otherwise, you'll call `Get()` on the future and send the result back to the client in the reply message.

- [x] `WorkflowCancelChildRequest:` You'll locate the parent workflow context and then use the `contextId` included in the message to locate the child context record.  If this doesn't exist, you should immediately send the reply with an **EntityNotExists** error.  Otherwise, you'll call the cancellation function and then remove the child context.

- [x] `WorkflowSetSignalHandlerRequest:` Lookup the workflow context and then create a function that accepts a single byte array as a parameter and returns a byte array and an error.  This function should send a `WorkflowSignalRequest` to the cadence-client passing the `SignalArgs`, wait for a `WorkflowSignalReply`, and return the result/error.  You'll call `SetQueryHandler(context,...), passing the `WorkflowSetSignalHandlerRequest.SignalName` and your function and then send a `WorkflowSetSignalHandlerReply` back to the client.

- [x] `WorkflowSetQueryHandlerRequest:` Lookup the workflow context and then create a function that accepts a single byte array as a parameter and returns a byte array and an error.  This function should send a `WorkflowQueryInvokeRequest` to the cadence-client passing the `QueryArgs`, wait for a `WorkflowQueryInvokeReply`, and return the result/error.  You'll call `SetQueryHandler(context,...), passing the `WorkflowSetSignalHandlerRequest.SignalName` and your function and then send a `WorkflowSetQueryHandlerReply` back to the client.

- [x] `ActivityExecuteLocalRequest:` Cadence supports the execution of local activities via the `ExecuteLocalActivity(context)` function.  Local activities are intended to be used running activities that run very quickly (up to a few seconds or so) with as little additional overhead as possible.  An example activity might be one that pings a server to see if its reachable which may take a few millseconds on a local network or perhaps 50-100ms if the server is out on the Internet.

  A normal cadence activity execution will need to registered with the Cadence cluster, the cluster will need to schedule the activity on a worker node, then run the activity, get the result, and then send the result back to the workflow that launched ther activity: a whole lot of overhead.
  
  Local activities aren't scheduled by the Cadence cluster.  Instead, they'll run locally on the worker hosting the workflow executing the activity.  Here's how we'll make this work for us:
  
  1. The .NET workflow app will call the cadence-client `Workflow.CallLocalActivity<TActivity>(byte[] args)` method, specifying the type implementing the activity.
  
  2. The `Workflow` class will internally register the activity type with a unique `long` ID (the `ActivityTypeId`.
  
  3. The `Workflow` class will send an `ActivityExecuteLocalRequest` message to the cadence-proxy, passing the activity arguments and the `ActivityTypeId`.
  
  4. cadence-proxy will create a function that accepts three parameters, a `Context`, a `long activityTypeId` and the arguments byte array.  This function will need to persist the context you received to the contexts table and then send the cadence-client an `ActivityInvokeLocalRequest` message, passing the current workflow context ID normally, the new workflow context ID as `ActivityContextId`, `activityTypeId` and the arguments. Then the function will wait for an `ActivityInvokeLocalReply`.  After receiving the reply, your function will remove the new context from your table and return the results, completing the operation from your perspective.

  5. cadence-client will receive the `ActivityInvokeLocalRequest`, route it to the calling `Workflow` instance, which will look up the activity type in a local table, using the `activityTypeId` you included in the message.  The workflow will then instantiate an instance of the activity type and invoke it, passing the workflow context ID and the arguments.  When the activity completes, the workflow will remove the activity type mapping, and then send the `ActivityExecuteLocalReply` with the result/error back to the cadence-proxy.
  
  Before you register the function, you'll need to deserialize `ActivityExecuteLocalRequest.Options` as a `LocalActivityOptions` value.  If this is not null then pass it to `WithLocalActivityOptions(Context, Options)` and use the context returned to actually register the function.  If `ActivityExecuteLocalRequest.Options` is null, then you'll just use the unchanged workflow context.

  **NOTE:** There are going to be two contexts for this operation, the original workflow context ID and the new activity context passed to your new function.  Both of these are important.

- [x] `ActivityRegisterRequest:` You'll call `RegisterActivityWithOptions()` much like you did for workflows.

- [x] `ActivityGetInfoRequest:` You'll call `activity.GetInfo()` and return the result as `ActivityGetInfoReply.Info`.  This is a JSON property that should map directly to the result.

- [x] `ActivityCompleteReply:` You'll call `client.CompleteActivity()`, passing the `TaskToken` and the result and error properties.

  
This completes the specification for this scenario and we're pretty close the being complete now.  Remaining work includes: logging, metrics, and a few listing operations.

**Comment by jefflill** [comment]
**DONE!**
